cpp/arcticdb/python/python_handlers.cpp

```diff
@@ -6,46 +6,204 @@
  */
 #include <python/python_handlers.hpp>
 #include <python/gil_lock.hpp>
+#include <arcticdb/codec/slice_data_sink.hpp>
+#include <arcticdb/codec/encoding_sizes.hpp>
 #include <arcticdb/codec/codec.hpp>
 
 namespace arcticdb {
 
-void EmptyHandler::handle_type(
-    const uint8_t *&input,
-    uint8_t *dest,
-    const VariantField &variant_field,
-    const entity::TypeDescriptor &,
-    size_t dest_bytes,
-    std::shared_ptr<BufferHolder>
-) {
-    ARCTICDB_SAMPLE(HandleEmpty, 0)
-    util::check(dest != nullptr, "Got null destination pointer");
-    const size_t num_rows = dest_bytes / get_type_size(DataType::EMPTYVAL);
-    static_assert(get_type_size(DataType::EMPTYVAL) == sizeof(PyObject *));
-    auto ptr_dest = reinterpret_cast<const PyObject **>(dest);
-    py::none none = {};
-    for (auto row = 0u; row < num_rows; ++row) {
-        none.inc_ref();
-        *ptr_dest++ = none.ptr();
+    /// @brief Generate numpy.dtype object from ArcticDB type descriptor
+    /// The dtype is used as type specifier for numpy arrays stored as column elements
+    /// @note There is special handling for ArcticDB's empty type
+    /// When numpy creates an empty array its type is float64. We want to mimic this because:
+    /// i) There is no equivalent to empty value
+    /// ii) We want input dataframes to be exact match of the output and that includes the type
+    [[nodiscard]] static inline py::dtype generate_python_dtype(const TypeDescriptor& td, stride_t type_byte_size) {
+        if(is_empty_type(td.data_type())) {
+            return py::dtype{"f8"};
+        }
+        return py::dtype{fmt::format("{}{:d}", get_dtype_specifier(td.data_type()), type_byte_size)};
+    }
+
+    /// @important This calls pybind's initialize array function which is NOT thread safe. Moreover, numpy arrays can
+    /// be created only by the thread holindg the GIL. In practice we can get away with allocating arrays only from
+    /// a sigle thread (enve if it's not the one holding the GIL). This, however, is not guranteed to work.
+    /// @todo Allocate numpy arrays only from the thread holding the GIL
```

Comment:
I think we have to sort out the GIL handling properly. This test,

```
    @pytest.mark.parametrize('n', range(1000))
    def test_append_empty_arrays_to_column(self, lmdb_version_store, array_type, n):
        df = pd.DataFrame({"col1": [np.array([1, 2, 3]).astype(array_type)]})
        lmdb_version_store.write("test_append_to_colum_with_empty_array", df)
        df_to_append = pd.DataFrame({"col1": [np.array([])]})
        lmdb_version_store.append("test_append_to_colum_with_empty_array", df_to_append)
        df_out = lmdb_version_store.read("test_append_to_colum_with_empty_array")
        df_target = pd.concat([df, df_to_append], ignore_index=True)
        assert_frame_equal(df_target, df_out.data)
```

that you added segfaults occasionally (hence the 1000 runs) in various places in Cython, indicating corruption of Python's datastructures.